### PR TITLE
Improved status command

### PR DESCRIPTION
**Pull request by patrickxb**
Here is a new `keybase status` command.  It outputs stuff like this:

```shell
Username:      kwejfkwef
Logged in:     yes

Device name:   home computer
Device ID:     dc268bf91b698658dbd4c1998ae4bc18
Device status: active

Local keybase keychain: unlocked
Session status:         kwejfkwef [salt only]

KBFS:
    status:    not running
    version:   
    log:       /Users/patrick/kbtest/Library/Logs/KeybaseDevel/keybase.kbfs.log

Service:
    status:    running
    version:   1.0.8-0
    log:       /Users/patrick/kbtest/Library/Logs/KeybaseDevel/keybase.service.log

Client:
    version:   1.0.8-0

Desktop app:
    status:    not running

Config path:        /Users/patrick/kbtest/Library/Application Support/KeybaseDevel/config.json
Default user:       kwejfkwef
Provisioned users:  jejjejje, kwkwwkwkkw, wjkwjekkw, 111
```

It can also output json and works logged out as well.

This runs now as `keybase nstatus`.  I haven't touched `keybase status` as I don't know what we should do with it...is the information that it provides more useful than `keybase device list`?

Also, the kbfs status of running/not running is only implemented for os x currently via @gabriel's `install` package.

r? @maxtaco @malgorithms 

cc @oconnor663 @cjb @gabriel

### CODE REVIEW

**Comment by malgorithms** [comment]
this output looks great to me. I don't know if anyone was using that old status command, so I personally would vote for replacing it, unless @maxtaco or someone else speaks up in defense.

**Comment by patrickxb** [comment]
I'd propose renaming the current `status` command to `keyfamily` and removing the first four lines.

**Comment by maxtaco** [comment]
I find the old version very useful and don't want to see it nuked.

**Comment by oconnor663** [comment]
`keyfamily` sounds great to me.

**Comment by malgorithms** [comment]
is it something users will ever use outside of our instruction? If not, then we should make sure it doesn't appear in help.

File: go/libkb/session.go [hunk 1]
```diff
@@ -252,6 +252,10 @@ func (s *Session) check() error {
 		return nil
 	}
 
+	if s.G().API == nil {
+		fmt.Printf("nil api???\n")
```
Comment (maxtaco) [inline go/libkb/session.go]: This seems like leftover debugging.

File: go/libkb/login_session.go [hunk 2]
```diff
@@ -38,11 +38,20 @@ func NewLoginSession(emailOrUsername string, g *GlobalContext) *LoginSession {
 func NewLoginSessionWithSalt(emailOrUsername string, salt []byte, g *GlobalContext) *LoginSession {
 	ls := NewLoginSession(emailOrUsername, g)
 	ls.salt = salt
+	// XXX are these right?  is this just so the salt can be retrieved?
 	ls.loaded = true
 	ls.cleared = true
 	return ls
 }
 
+func (s *LoginSession) Status() string {
```
Comment (maxtaco) [inline go/libkb/login_session.go]: can you return a struct here rather than a string?

File: go/libkb/util.go [hunk 3]
```diff
@@ -445,3 +445,12 @@ func DigestForFileAtPath(path string) (digest string, err error) {
 	digest = hex.EncodeToString(hasher.Sum(nil))
 	return
 }
+
+// BoolString maps a bool to true and false strings (like "yes"
+// and "no").  For example:  BoolString(b, "yes", "no").
+func BoolString(b bool, t, f string) string {
```
Comment (maxtaco) [inline go/libkb/util.go]: if returning a struct above, wouldn't need this

File: protocol/avdl/config.avdl [hunk 4]
```diff
@@ -13,6 +13,21 @@ protocol config {
...
   GetCurrentStatusRes getCurrentStatus(int sessionID);
 
+  record ExtendedStatus {
+    boolean standalone;
+    boolean passphraseStreamCached;
+    DeviceID deviceID;
+    string deviceName;
+    string deviceStatus;
+    string logDir;
+    boolean desktopUIConnected;
```
Comment (maxtaco) [inline protocol/avdl/config.avdl]: eventually we should also show: pid, path, and process name/type; not just if is connected

It's slightly weird `ExtendedStatus` isn't a superset of `Status`.  The overlap is actually quite small.

File: go/client/cmd_nstatus.go [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+		return err
+	}
+	dui := c.G().UI.GetDumbOutputUI()
+	_, err = dui.Printf(string(b) + "\n")
+	return err
+}
+
+func (c *CmdNStatus) outputTerminal(status *fstatus) error {
+	dui := c.G().UI.GetDumbOutputUI()
+	dui.Printf("Username:      %s\n", status.Username)
```
Comment (maxtaco) [inline go/client/cmd_nstatus.go]: does it make sense to use our `Tablify` function?

File: protocol/avdl/config.avdl (hunk 4 above)
Comment (maxtaco) [inline protocol/avdl/config.avdl]: Ah, I see, it's meant for you to call them separately

**Comment by maxtaco** [comment]
Can the" "hello" RPC also say where the log is?  That seems very reasonable too.

File: protocol/avdl/config.avdl (hunk 4 above)
Comment (patrickxb) [inline protocol/avdl/config.avdl]: There are a lot of callers of GetCurrentStatus now...I didn't want to mess with the protocol.

**Comment by patrickxb** [comment]
Yes, but currently it's stdout for service and kbfs.  It's only the process manager (like launchd) that sends stdout to a file.  So kbfs doesn't really know where the log file is.

**Comment by maxtaco** [comment]
Let's call the old command `keybase dump-keyfamily`?

**Comment by maxtaco** [comment]
(Got it about the log file location, good point)

File: go/libkb/util.go (hunk 3 above)
Comment (patrickxb) [inline go/libkb/util.go]: it's used in cmd_nstatus.go too.

Comment (maxtaco) [inline go/libkb/util.go]: ok1

Comment (maxtaco) [inline go/libkb/util.go]: Also, `"%v"` outputs true/false for bools, which might be ok.  But this is all very nitty-

Comment (patrickxb) [inline go/libkb/util.go]: It was.  I moved it here for this Status() function.

Comment (patrickxb) [inline go/libkb/util.go]: yes, I know about %v and bool, but I liked yes/no and unlocked/locked.

File: go/client/cmd_nstatus.go (hunk 5 above)
Comment (patrickxb) [inline go/client/cmd_nstatus.go]: you mean the UI-related helper in libkb :smile: ?

I tried with tabs, but the data started to get far from the headings.

File: go/libkb/login_session.go (hunk 2 above)
Comment (patrickxb) [inline go/libkb/login_session.go]: A struct for the status is going to end up just copying half the fields of this type, but ok...

**Comment by patrickxb** [comment]
Ok, old `status` is now `dump-keyfamily`.
`nstatus` is now `status`.
And minor fixes from PR done.

PTAL

**Comment by patrickxb** [comment]
And `dump-keyfamily` doesn't show up in `keybase help`

**Comment by maxtaco** [comment]
:+1: but while you are there, can you contextify `dump-keyfamily` and not use `GlobUI`? 

At this point, we're almost done!

**Pull request merged**